hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/update/server/SCMCRLUpdateHandler.java

```
@@ -0,0 +1,151 @@
+    return TYPE;
+  }
+
+  @Override
+  public void handleClientRequest(SCMUpdateServiceProtos.UpdateRequest request,
+      SCMUpdateClientInfo clientInfo) {
+    SCMUpdateServiceProtos.CRLUpdateRequest updateStatusRequest =
+        request.getCrlUpdateRequest();
+    long clientCrlId = updateStatusRequest.getReceivedCrlId();
+
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Client {} updateStatus \nclientCrlId {}  \npendingCrls {}",
+          clientInfo.getClientId(), clientCrlId,
+          updateStatusRequest.getPendingCrlIdsList().toString());
+    }
+
+    CRLClientInfo crlClientInfo;
+    if (!clients.containsKey(clientInfo.getClientId())) {
+      crlClientInfo = new CRLClientInfo(clientInfo);
+      clients.put(clientInfo.getClientId(), crlClientInfo);
+    } else {
+      crlClientInfo = clients.get(clientInfo.getClientId());
+    }
+
+    crlClientInfo.setPendingCrlIds(
+        request.getCrlUpdateRequest().getPendingCrlIdsList());
+    crlClientInfo.setReceivedCrlId(
+        request.getCrlUpdateRequest().getReceivedCrlId());
+
+    sendCrlUpdateToClient(crlClientInfo);
+  }
+
+  @Override
+  public void onUpdate() {
+    LOG.debug("Update due to certificate revoke");
+    // server crl id is usually > client crl id when this is invoked.
+    clients.values().forEach(client -> {
+      sendCrlUpdateToClient(client);
+    });
+  }
+
+  @Override
+  public void onRemoveClient(SCMUpdateClientInfo clientInfo) {
+    clients.remove(clientInfo.getClientId());
+  }
+
+  private void sendCrlUpdateToClient(CRLClientInfo client) {
+    long clientCrlId = client.getReceivedCrlId();
+    long serverCrlId = crlStore.getLatestCrlId();
+
+    if (clientCrlId >= serverCrlId) {
+      return;
+    }
+
+    LOG.debug("## Server: clientCrlId {} serverCrlId {}",
+        clientCrlId, serverCrlId);
+
+    long nextCrlId = clientCrlId + 1;
+    try {
+      CRLInfo crlInfo = null;
+      while (crlInfo == null && nextCrlId <= serverCrlId) {
```

Why do we need the null check here? This can be replaced with if
```suggestion
      if (nextCrlId <= serverCrlId) {
```

---

This is to handle potential holes in the crlId sequence (might come from unit test). Server can still send the next crlId to client even though there is a gap. Correctness wise, I think we should throw an exception here if server has a gap in the crl sequence id.